### DISCUSSION TITLE
Would it be a good idea to bring zlib-ng to the attention of the Linux Kernel folks?

### DISCUSSION THREAD

**Discussion by ms178** [zlib-ng/zlib-ng#1291] [General]
A recent GCC bug lead to an effort to modernize the zlib library used in the Linux Kernel (see: http://lkml.iu.edu/hypermail/linux/kernel/2105.0/01493.html). I am not subscribed to the list and my name doesn't carry any weight anyway, so maybe one of the interested parties here (Google? ARM? IBM?) could pitch the idea of using zlib-ng as a drop-in replacement and enjoy some of the benefits which come with it? 

Of course only if it would make sense.

**Comment by nmoinvaz** [comment]
Thanks for letting us know, I will try and reach out.

**Comment by nmoinvaz** [comment]
Linus responded regarding the zlib changes that it was mostly related to a compiler issue. He also mentioned that there are some specific modifications made to their zlib implementation that are necessary for the kernel, namely no dynamic memory allocations using `malloc`.

> The problem with the kernel side of zlib is that we can't just act as if it's just another regular library: the kernel tends to have some restrictions on what we can do _anyway_, but when it comes to zlib (and some other code), those restrictions are even worse than usual because it's used fairly early during boot for uncompressing early data.

Anybody can view their zlib implementation that is in the kernel and it might benefit us to take a look at their changes:

https://github.com/torvalds/linux/tree/master/lib/zlib_deflate
https://github.com/torvalds/linux/tree/master/lib/zlib_inflate

In deflate.c you can see that they use a `deflate_workspace` structure.
